File: Graph/KeysAndRooms/key.py
from typing import List

class Solution:
    def doDFS(self,grid,visited,roomNum):
        if not visited[roomNum]:
            visited[roomNum] = True
        for neighbour in grid[roomNum]:
            if not visited[neighbour]: self.doDFS(grid,visited,neighbour)
        return
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        if n == 0: return True
        visited = [False]*n
        self.doDFS(rooms,visited,0)
        for i in range(n):
            if not visited[i]: return False
        return True

File: Graph/KeysAndRooms/key.cpp
#include <vector>
using namespace std;

class Solution {
public:
    void doDFS(vector<vector<int>>& rooms, vector<bool>& visited, int roomNum) {
        if (visited[roomNum])
            return;

        visited[roomNum] = true;

        for (int neighbour : rooms[roomNum]) {
            if (!visited[neighbour]) {
                doDFS(rooms, visited, neighbour);
            }
        }
    }

    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        if (n == 0)
            return true;

        vector<bool> visited(n, false);

        doDFS(rooms, visited, 0);

        for (int i = 0; i < n; i++) {
            if (!visited[i])
                return false;
        }

        return true;
    }
};